csharp_dotnetcore/14.NLP-With-Dispatch/NlpDispatch/NlpDispatchBot.cs

```diff
@@ -0,0 +1,182 @@
+            }
+
+            if (!_services.LuisServices.ContainsKey(HomeAutomationLuisKey))
+            {
+                throw new System.ArgumentException($"Invalid configuration.  Please check your '.bot' file for a Luis service named '{HomeAutomationLuisKey}'.");
+            }
+
+            if (!_services.LuisServices.ContainsKey(WeatherLuisKey))
+            {
+                throw new System.ArgumentException($"Invalid configuration.  Please check your '.bot' file for a Luis service named '{WeatherLuisKey}'.");
+            }
+        }
+
+        /// <summary>
+        /// Every Conversation turn for our NLP Dispatch Bot will call this method.
+        /// There are no dialogs used, since it's "single turn" processing, meaning a single
+        /// request and response, with no stateful conversation.
+        /// </summary>
+        /// <param name="context">A <see cref="ITurnContext"/> containing all the data needed
+        /// for processing this conversation turn. </param>
+        /// <param name="cancellationToken">(Optional) A <see cref="CancellationToken"/> that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        public async Task OnTurnAsync(ITurnContext context, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (context.Activity.Type == ActivityTypes.Message && !context.Responded)
+            {
+                // Get the intent recognition result
+                var recognizerResult = await _services.LuisServices[DispatchKey].RecognizeAsync(context, cancellationToken);
+                var topIntent = recognizerResult?.GetTopScoringIntent();
+
+                if (topIntent == null)
+                {
+                    await context.SendActivityAsync("Unable to get the top intent.");
+                }
+                else
+                {
+                    if (topIntent.Value.score < 0.3)
+                    {
+                        await context.SendActivityAsync("I'm not very sure what you want but will try to send your request.");
+                    }
+
+                    await DispatchToTopIntentAsync(context, topIntent);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Depending on the intent from Dispatch, routes to the right Luis model or Qna service.
+        /// </summary>
+        private async Task DispatchToTopIntentAsync(ITurnContext context, (string intent, double score)? topIntent)
+        {
+            const string homeAutomationDispatchKey = "l_homeautomation-LUIS";
+            const string weatherDispatchKey = "l_weather-LUIS";
+            const string noneDispatchKey = "None";
+            const string qnaDispatchKey = "q_sample_qna";
+
+            switch (topIntent.Value.intent.ToLowerInvariant())
+            {
+                case homeAutomationDispatchKey:
```

Comment:
You should use LUISGen to generate strong typed C# classes for all LUIS models. This wasnt an option for me in JS for the dispatch sample but we should totally use that and recommend it forC#

Reply:
+1

---
In reply to: [215112933](https://github.com/Microsoft/BotBuilder-Samples/pull/319#discussion_r215112933) [](ancestors = 215112933)

Reply:
I'll sync up with Dave on this so we add it to samples.